src/psbt.c

```diff
@@ -2192,347 +2388,236 @@ int wally_psbt_to_base64(
-    result->num_outputs = psbts[0].num_outputs;
+    COMBINE_BYTES(output, redeem_script);
+    COMBINE_BYTES(output, witness_script);
+
+#ifdef BUILD_ELEMENTS
+    COMBINE_BYTES(output, blinding_pubkey);
+    COMBINE_BYTES(output, value_commitment);
+    COMBINE_BYTES(output, value_blinder);
+    COMBINE_BYTES(output, asset_commitment);
+    COMBINE_BYTES(output, asset_blinder);
+    COMBINE_BYTES(output, nonce_commitment);
+    COMBINE_BYTES(output, range_proof);
+    COMBINE_BYTES(output, surjection_proof);
+#endif
+    return WALLY_OK;
+}
+#undef COMBINE_BYTES
 
-    for (i = 0; i < psbts_len; ++i) {
-        unsigned char txid[SHA256_LEN];
+int wally_psbt_combine(struct wally_psbt *psbt, const struct wally_psbt *src)
+{
+    unsigned char txid[WALLY_TXHASH_LEN], src_txid[WALLY_TXHASH_LEN];
+    size_t i;
+    int ret;
 
-        /* Compare the txids */
-        if ((ret = get_txid(psbts[i].tx, txid, SHA256_LEN)) != WALLY_OK) {
-            goto fail;
-        }
-        if (memcmp(global_txid, txid, SHA256_LEN) != 0) {
-            ret = WALLY_EINVAL;
-            goto fail;
-        }
+    if (!psbt || !psbt->tx || !src || !src->tx)
+        return WALLY_EINVAL;
 
-        /* Now start merging */
-        for (j = 0; j < result->num_inputs; ++j) {
-            if ((ret = merge_input_into(&result->inputs[j], &psbts[i].inputs[j])) != WALLY_OK) {
-                goto fail;
-            }
-        }
-        for (j = 0; j < result->num_outputs; ++j) {
-            if ((ret = merge_output_into(&result->outputs[j], &psbts[i].outputs[j])) != WALLY_OK) {
-                goto fail;
-            }
-        }
+    ret = wally_tx_get_txid(psbt->tx, txid, sizeof(txid));
+    if (ret == WALLY_OK)
+        ret = wally_tx_get_txid(src->tx, src_txid, sizeof(src_txid));
 
-        if (psbts[i].unknowns) {
-            if (!result->unknowns) {
-                if ((ret = wally_unknowns_map_init_alloc(psbts[i].unknowns->items_allocation_len, &result->unknowns)) != WALLY_OK) {
-                    goto fail;
-                }
-            }
+    if (ret == WALLY_OK && memcmp(txid, src_txid, sizeof(txid)) != 0)
+        ret = WALLY_EINVAL; /* Transactions don't match */
 
-            if ((ret = merge_unknowns_into(result->unknowns, psbts[i].unknowns)) != WALLY_OK) {
-                return ret;
-            }
-        }
-    }
+    for (i = 0; ret == WALLY_OK && i < psbt->num_inputs; ++i)
+        ret = combine_inputs(&psbt->inputs[i], &src->inputs[i]);
 
-    *output = result;
-    return WALLY_OK;
+    for (i = 0; ret == WALLY_OK && i < psbt->num_outputs; ++i)
+        ret = combine_outputs(&psbt->outputs[i], &src->outputs[i]);
+
+    if (ret == WALLY_OK)
+        ret = combine_unknowns(&psbt->unknowns, src->unknowns);
 
-fail:
-    wally_psbt_free(result);
     return ret;
 }
 
-int wally_sign_psbt(
-    struct wally_psbt *psbt,
-    const unsigned char *key,
-    size_t key_len)
+int wally_psbt_sign(struct wally_psbt *psbt,
```

Comment:
I wonder if we should add a `flags` parameter here. Currently it signs with `EC_FLAG_ECDSA | EC_FLAG_GRIND_R`, but users may want to use different flags. OTOH if `EC_FLAG_SCHNORR` should be used it can probably be deduced from the signed scripts and `EC_FLAG_RECOVERABLE` should not be accepted, so it may be ok as it is.

Reply:
I assume we can tell ECDSA from Schnorr by the witness/scripts, but we should probably accept at least the grinding flag (super low power devices may opt not to grind if signing takes too long, for example). And there may be other relevant flags in the future like alternate nonce functions or something relating to signature aggregation. I need to think on this a bit.

However this PR hasn't addressed the signing code yet, that will probably be a separate PR, looks to me like the current impl can be fairly easily made to crash. This PR can't fix everything that needs doing though, or it will end up well over its current ~200 commits :(

Reply:
I've added the ability to control grinding (and extend with other flags in the future).